mantidproject/mantid
# Remove Visual Studio compiler warnings

**Description of work.**

There are two new compiler warnings when Mantid is compiled in Visual Studio ~v16.7 or later:
a) C4996. Warning that std::shared_ptr::unique() has been deprecated
b) C4250. Warning that Algorithm and all the derived classes inherit various methods "by dominance"

As well as indicating problems in the code, the large number of these warnings were slowing down the compilation in Visual Studio.
This PR suppresses them both.

a) has been suppressed by simply adding a suppression to the MSVCSetup.cmake file. This is a bit of a stop gap while an alternative solution is investigated. See https://github.com/mantidproject/mantid/issues/30266

Suppressing b) involved a change to the Algorithm class hierarchy to remove a diamond at the top (a diagram of the original hierarchy is shown on the linked issue). PropertyManagerOwner has been added to the Algorithm class by composition instead of by inheritance which removes the diamond. This involved adding various forwarding functions to the Algorithm class. This leaves open the possibility of retiring PropertyManagerOwner at some point and moving its minimal functionality in Algorithm. There were a couple of places where PropertyManagerOwner class was unnecessarily referred to so I've removed these to make it easier to remove it in future.

The code base contains a lot of lines like this:
```
IAlgorithm_sptr gatherAlg = createChildAlgorithm("GatherWorkspaces");
```

The IAlgorithm_sptr type has been replaced with "auto" to provide some extra flexibility around the arrangement of the class\interface hierarchy in future. The exact form of the Algorithm interfaces was changed a couple of times before landing on the final hierarchy in this PR and it was helpful to have the specific IAlgorithm_sptr type removed while doing this work. I've left these "auto" changes because it's possible this could be changed again in future.

**To test:**

Build Framework in Visual Studio and check you don't get hundreds\thousands of C4250 and C4996 compiler warnings

Fixes #30271.

*This does not require release notes* because **no functional change**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
